File: 1185-find-in-mountain-array/1185-find-in-mountain-array.cpp
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */

class Solution {
public:
    int findpeak( MountainArray &mountainArr){
        int start = 0;
        int end = mountainArr.length()-1;
        int mid = start + (end-start)/2;
        while(start<end){
            if(mountainArr.get(mid+1)>mountainArr.get(mid)){
                start = mid+1;
            }else{
                end= mid;
            }
            mid = start + (end-start)/2;
        }
        return start;
    }
    int binarysearch(int start, int end ,int target,bool asc,MountainArray &mountainArr){
        int mid = start + (end-start)/2;
        while(start <= end){
            if(asc==true){
                if(mountainArr.get(mid)==target){
                return mid;
            }else if(mountainArr.get(mid)>target){
                end = mid-1;
            }else{
                start = mid+1;  
            }
            }else{
                if(mountainArr.get(mid)==target){
                return mid;
            }else if(mountainArr.get(mid)>target){
                start = mid+1;
            }else{
                end = mid-1;  
            }
            }
            mid = start + (end-start)/2;
        }
        return -1;

    }
    int findInMountainArray(int target, MountainArray &mountainArr) {
        int index = findpeak(mountainArr);
        int ans = binarysearch(0,index,target,true,mountainArr);
        if(ans == -1){
            ans =  binarysearch(index,mountainArr.length()-1,target,false,mountainArr);
        }
        return ans;
    }
};

File: 1185-find-in-mountain-array/1185-find-in-mountain-array.java
/**
 * // This is MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface MountainArray {
 *     public int get(int index) {}
 *     public int length() {}
 * }
 */
 
class Solution {
    private int pivotindex(MountainArray arr){
        int start = 0;
        int end = arr.length()-1;
        while(start<end){
            int mid = start + (end - start)/2;
            if(arr.get(mid)<arr.get(mid+1)){
                start = mid+1;
            }else{
                end = mid;
            }
        }
        return start;
    }
    private int binarysearch(MountainArray arr, int target, boolean asc,int start,int end){
        while(start<=end){
            int mid = start + (end-start)/2;
            int ele = arr.get(mid);
            if(ele == target) return mid;
            if(asc){
                if(ele>target){
                    end = mid-1;
                }else{
                    start = mid+1;
                }
            }else{
                if(ele>target){
                    start = mid+1;
                }else{
                    end = mid-1;
                }            }
        }
        return -1;
    }
    public int findInMountainArray(int target, MountainArray mountainArr) {
       int peak = pivotindex(mountainArr);
       int index = binarysearch(mountainArr,target,true,0,peak);
       if(index!=-1) return index;
        return binarysearch(mountainArr,target,false,peak+1,mountainArr.length()-1);
    }
}